src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java

```diff
@@ -16,43 +16,196 @@
+                    ));
+
+            // Locate the uploaded GAX file
+            final List<Path> gaxFiles = uploadedResources.entrySet().stream()
+                    .filter(entry -> FilenameUtils.isExtension(entry.getKey(), "gax"))
+                    .map(Map.Entry::getValue)
+                    .collect(Collectors.toList());
+            if (gaxFiles.size() != 1) {
+                throw new ValidationException("GI Archive must contain one and only one .gax GI definition XML file");
+            }
+            final Path gaxFile = gaxFiles.get(0);
+
+            // Save GAX / GI definition in item.json
+            try (final InputStream gaxStream = Files.newInputStream(gaxFile)) {
+                contentModel.setGiDefinition(IOUtils.toString(gaxStream, StandardCharsets.UTF_8).trim());
+            }
+
+            // Validate and upload GAX-referenced resources
+            final List<String> referencedFilenames = AttachmentSelector.getReferencedGaxFileNames(gaxFile.toFile());
+            final List<Attachment> referencedResources = referencedFilenames.stream()
+                    .map(referencedFilename -> {
+                        if (!uploadedResources.containsKey(referencedFilename)) {
+                            throw new ValidationException(format("GAX references missing resource: %s", referencedFilename));
+                        }
+                        return uploadedResources.get(referencedFilename);
+                    })
+                    .map(resource -> uploadCoreFile(
+                            archiveUpload.getItemId(),
+                            userTransaction.getSection(),
+                            resource.getFileName().toString(),
+                            resource
+                    ))
+                    .collect(Collectors.toList());
+
+            // Include translated GAX resource overrides
+            // TODO These should probably be represented by the UI and model as translation resource overrides
+            //  rather than reply on zip-contents file-naming conventions.
+            final Set<String> referencedFilenameBases = referencedFilenames.stream()
+                    .map(FilenameUtils::removeExtension)
+                    .collect(Collectors.toSet());
+
+            final List<Attachment> translationResources = uploadedResources.entrySet().stream()
+                    .filter(entry -> {
+                        final String uploadedFilename = removeExtension(entry.getKey());
+                        final Matcher matcher = TRANSLATED_RESOURCE_FILENAME.matcher(uploadedFilename);
+                        if (!matcher.matches()) {
+                            return false;
+                        }
+
+                        final String referencedFilename = matcher.group(1);
+                        return referencedFilenameBases.contains(referencedFilename);
+                    })
+                    .map(entry -> uploadCoreFile(
```

Comment:
side effects should never occur within a map call, can you create the objects in a map call and then for-each that for the ops?

Reply:
I could, but then my forEach needs to append the response from `uploadCoreFile(...)` to an instantiated list. ...which is basically a map with an additional data structure... thoughts?

Reply:
Hm. maybe just add a comment here to that affect for now

Reply:
Maybe it'll look cleaner after I break some of this out into submethods

Reply:
I'm still uploading in a `.map` after the reorg but it's the only thing happening in the map, so hopefully it reads a little cleaner.